We Continue triggering Alerts for Deleted Cards

### Describe the bug

In some cases when a card is archived the associated alerts trigger aren't being deleted/archived. During the notification send process Metabase still attempts to deliver them (which fails with an error since the card doesn't exist).

### To Reproduce

1. Add a card to a collection
2. Set up an alert on the card that executed every minute
3. After the alert has been created move the card to a dashboard
4. Move the dashboard, containing the card, to the trash and then permanently delete it
5. Watch the logs and see that every minute the notification will attempt to send and fail

### Expected behavior

When we archive cards that aren't stored in dashboards the associated alerts are removed entirely. I assume the behavior should be the same in this case.

### Logs

```
2025-12-24 00:18:02,050 ERROR notification.send :: Error in notification worker {mb-card_id=, mb-notification-id=13, mb-	notification_id=13, mb-payload_type=:notification/card, mb-quartz-job-type=SendNotification, mb-subscription-id=13}

java.lang.AssertionError: Assert failed: (pos-int? card-id)

at metabase.query_processor.card$process_query_for_card.invokeStatic(card.clj:297)

at metabase.query_processor.card$process_query_for_card.doInvoke(card.clj:297)

at clojure.lang.RestFn.invoke(RestFn.java:829)

at metabase.notification.payload.execute$execute_card$fn__112533.invoke(execute.clj:290)

at metabase.request.session$do_with_current_user$fn__85481.invoke(session.clj:60)

at metabase.request.session$do_with_current_user$do_with_user_local_values__85476.invoke(session.clj:56)

at metabase.request.session$do_with_current_user.invokeStatic(session.clj:57)

at metabase.request.session$do_with_current_user.invoke(session.clj:38)

at metabase.notification.payload.execute$execute_card.invokeStatic(execute.clj:288)

at metabase.notification.payload.execute$execute_card.invoke(execute.clj:284)

at metabase.notification.payload.impl.card$fn__145481$f__15665__auto____145483$fn__145485$fn__145486.invoke(card.clj:19)

at metabase.util.log$with_exception_context_fn.invokeStatic(log.clj:303)

at metabase.util.log$with_exception_context_fn.invoke(log.clj:295)

at metabase.notification.payload.impl.card$fn__145481$f__15665__auto____145483$fn__145485.invoke(card.clj:17)

at metabase.util.log$with_thread_context_fn.invokeStatic(log.clj:85)

at metabase.util.log$with_thread_context_fn.invoke(log.clj:71)

at metabase.notification.payload.impl.card$fn__145481$f__15665__auto____145483.invoke(card.clj:17)

at clojure.lang.MultiFn.invoke(MultiFn.java:229)

at metabase.notification.payload.core$notification_payload.invokeStatic(core.clj:151)

at metabase.notification.payload.core$notification_payload.invoke(core.clj:146)

at metabase.notification.send$send_notification_sync_BANG_$fn__114916$fn__114917$fn__114918$fn__114919$fn__114924.invoke(send.clj:148)

at metabase.task_history.models.task_history$do_with_task_history.invokeStatic(task_history.clj:157)

at metabase.task_history.models.task_history$do_with_task_history.invoke(task_history.clj:145)

at metabase.notification.send$send_notification_sync_BANG_$fn__114916$fn__114917$fn__114918$fn__114919.invoke(send.clj:145)

at metabase.notification.send$send_notification_sync_BANG_$fn__114916$fn__114917$fn__114918.invoke(send.clj:140)

at metabase.util$do_with_timer_ms.invokeStatic(util.cljc:1305)

at metabase.util$do_with_timer_ms.invoke(util.cljc:1300)

at metabase.notification.send$send_notification_sync_BANG_$fn__114916$fn__114917.invoke(send.clj:136)

at metabase.util.log$with_exception_context_fn.invokeStatic(log.clj:303)

at metabase.util.log$with_exception_context_fn.invoke(log.clj:295)

at metabase.notification.send$send_notification_sync_BANG_$fn__114916.invoke(send.clj:134)

at metabase.util.log$with_thread_context_fn.invokeStatic(log.clj:85)

at metabase.util.log$with_thread_context_fn.invoke(log.clj:71)

at metabase.notification.send$send_notification_sync_BANG_.invokeStatic(send.clj:134)

at metabase.notification.send$send_notification_sync_BANG_.invoke(send.clj:131)

at metabase.notification.send$create_notification_dispatcher$start_worker_BANG___115063$fn__115064$fn__115065$fn__115066$fn__115067.invoke(send.clj:365)

at metabase.util.log$with_exception_context_fn.invokeStatic(log.clj:303)

at metabase.util.log$with_exception_context_fn.invoke(log.clj:295)

at metabase.notification.send$create_notification_dispatcher$start_worker_BANG___115063$fn__115064$fn__115065$fn__115066.invoke(send.clj:364)

at metabase.util.log$with_thread_context_fn.invokeStatic(log.clj:85)
```

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "h2"
    ],
    "run-mode": "prod",
    "plan-alias": "enterprise-self-hosted",
    "version": {
      "date": "2025-12-23",
      "tag": "v1.57.7.8",
      "hash": "338e01e"
    },
    "settings": {
      "report-timezone": null
    },
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "15.1 (Debian 15.1-1.pgdg110+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.7"
      }
    }
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "21.0.9+10-LTS",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "21.0.9",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "21.0.9+10-LTS",
    "os.name": "Linux",
    "os.version": "5.15.167.4-microsoft-standard-WSL2",
    "user.language": "en",
    "user.timezone": "GMT"
  }
}
```

### Severity

annoying

### Additional context

As far as I can tell with local reproductions this doesn't prevent any other notifications from sending. But it does clog up the tasks and logs with send failures in some cases (which is making it difficult to uncover other notification issues).

--- a/.clj-kondo/config/modules/config.edn
+++ b/.clj-kondo/config/modules/config.edn
@@ -1640,6 +1640,7 @@
                     :model/Field
                     :model/ModerationReview
                     :model/NativeQuerySnippet
+                    :model/Notification
                     :model/Revision
                     :model/Table
                     :model/User}}

--- a/src/metabase/notification/send.clj
+++ b/src/metabase/notification/send.clj
@@ -140,6 +140,13 @@
       (try
         (log/info "Sending")
         (prometheus/inc! :metabase-notification/concurrent-tasks)
+        ;; Guard against orphaned card notifications whose payload record was cascade-deleted
+        (when (and (= :notification/card payload_type)
+                   (not (t2/exists? :model/NotificationCard (:payload_id notification-info))))
+          (log/warnf "Payload for notification %d no longer exists, deleting" id)
+          (t2/delete! :model/Notification id)
+          (throw (ex-info "Card no longer exists, notification deleted"
+                          {:notification-id id})))
         (let [hydrated-notification (hydrate-notification notification-info)
               handlers              (:handlers hydrated-notification)]
           (task-history/with-task-history {:task          "notification-send"

--- a/src/metabase/notification/send.clj
+++ b/src/metabase/notification/send.clj
@@ -140,13 +140,16 @@
       (try
         (log/info "Sending")
         (prometheus/inc! :metabase-notification/concurrent-tasks)
-        ;; Guard against orphaned card notifications whose payload record was cascade-deleted
-        (when (and (= :notification/card payload_type)
-                   (not (t2/exists? :model/NotificationCard (:payload_id notification-info))))
-          (log/warnf "Payload for notification %d no longer exists, deleting" id)
-          (t2/delete! :model/Notification id)
-          (throw (ex-info "Card no longer exists, notification deleted"
-                          {:notification-id id})))
+        ;; Guard against orphaned card notifications whose payload record was cascade-deleted.
+        ;; Only applies to persisted notifications (with a non-nil :payload_id); Pulse-converted
+        ;; notifications pass through :payload instead and have no :payload_id.
+        (when-let [payload-id (when (= :notification/card payload_type)
+                                (:payload_id notification-info))]
+          (when-not (t2/exists? :model/NotificationCard payload-id)
+            (log/warnf "Payload for notification %d no longer exists, deleting" id)
+            (t2/delete! :model/Notification id)
+            (throw (ex-info "Card no longer exists, notification deleted"
+                            {:notification-id id}))))
         (let [hydrated-notification (hydrate-notification notification-info)
               handlers              (:handlers hydrated-notification)]
           (task-history/with-task-history {:task          "notification-send"

--- a/src/metabase/queries/models/card.clj
+++ b/src/metabase/queries/models/card.clj
@@ -847,7 +847,16 @@
   ;; delete any ParameterCard linked to this card
   (t2/delete! :model/ParameterCard :card_id id)
   (t2/delete! :model/ModerationReview :moderated_item_type "card", :moderated_item_id id)
-  (t2/delete! :model/Revision :model "Card", :model_id id))
+  (t2/delete! :model/Revision :model "Card", :model_id id)
+  ;; delete any card-type notifications for this card — must materialize IDs first because
+  ;; Notification's before-delete deletes the NotificationCard, which would make a subquery
+  ;; return empty by the time the actual DELETE executes.
+  (when-let [notification-ids (seq (t2/select-pks-set :model/Notification
+                                                      :payload_type :notification/card
+                                                      :payload_id [:in {:select [:id]
+                                                                        :from   [:notification_card]
+                                                                        :where  [:= :card_id id]}]))]
+    (t2/delete! :model/Notification :id [:in notification-ids])))
 
 (defmethod serdes/hash-fields :model/Card
   [_card]

--- a/test/metabase/notification/models_test.clj
+++ b/test/metabase/notification/models_test.clj
@@ -81,23 +81,32 @@
           (is (not (t2/exists? :model/NotificationCard notification-card-id))))))))
 
 (deftest delete-card-cleans-up-notifications-test
-  (testing "deleting a card should remove all associated active notifications"
+  (testing "deleting a card should remove all associated notifications, both active and inactive"
     (notification.tu/with-notification-cleanup!
       (mt/with-temp [:model/Card {card-id :id} {:name          "alert cleanup test card"
                                                 :dataset_query (mt/mbql-query products {:aggregation [[:count]]})}]
-        (let [notification (models.notification/create-notification!
-                            {:payload_type :notification/card
-                             :payload      {:card_id card-id}
-                             :creator_id   (mt/user->id :crowberto)}
-                            []
-                            [])]
-          (testing "sanity: notification and notification card exist"
-            (is (t2/exists? :model/Notification (:id notification)))
-            (is (t2/exists? :model/NotificationCard :card_id card-id)))
+        (let [active-notification   (models.notification/create-notification!
+                                     {:payload_type :notification/card
+                                      :payload      {:card_id card-id}
+                                      :creator_id   (mt/user->id :crowberto)}
+                                     []
+                                     [])
+              inactive-notification (models.notification/create-notification!
+                                     {:payload_type :notification/card
+                                      :payload      {:card_id card-id}
+                                      :active       false
+                                      :creator_id   (mt/user->id :crowberto)}
+                                     []
+                                     [])]
+          (testing "sanity: both notifications and notification cards exist"
+            (is (t2/exists? :model/Notification (:id active-notification)))
+            (is (t2/exists? :model/Notification (:id inactive-notification)))
+            (is (= 2 (t2/count :model/NotificationCard :card_id card-id))))
           (t2/delete! :model/Card card-id)
-          (testing "notification should be removed when the card is deleted"
-            (is (not (t2/exists? :model/Notification (:id notification)))))
-          (testing "notification card payload should also be removed"
+          (testing "both notifications should be removed when the card is deleted"
+            (is (not (t2/exists? :model/Notification (:id active-notification))))
+            (is (not (t2/exists? :model/Notification (:id inactive-notification)))))
+          (testing "notification card payloads should also be removed"
             (is (not (t2/exists? :model/NotificationCard :card_id card-id)))))))))
 
 (deftest notification-subscription-type-test

--- a/test/metabase/notification/models_test.clj
+++ b/test/metabase/notification/models_test.clj
@@ -80,6 +80,26 @@
           (t2/delete! :model/Notification (:id notification))
           (is (not (t2/exists? :model/NotificationCard notification-card-id))))))))
 
+(deftest delete-card-cleans-up-notifications-test
+  (testing "deleting a card should remove all associated active notifications"
+    (notification.tu/with-notification-cleanup!
+      (mt/with-temp [:model/Card {card-id :id} {:name          "alert cleanup test card"
+                                                :dataset_query (mt/mbql-query products {:aggregation [[:count]]})}]
+        (let [notification (models.notification/create-notification!
+                            {:payload_type :notification/card
+                             :payload      {:card_id card-id}
+                             :creator_id   (mt/user->id :crowberto)}
+                            []
+                            [])]
+          (testing "sanity: notification and notification card exist"
+            (is (t2/exists? :model/Notification (:id notification)))
+            (is (t2/exists? :model/NotificationCard :card_id card-id)))
+          (t2/delete! :model/Card card-id)
+          (testing "notification should be removed when the card is deleted"
+            (is (not (t2/exists? :model/Notification (:id notification)))))
+          (testing "notification card payload should also be removed"
+            (is (not (t2/exists? :model/NotificationCard :card_id card-id)))))))))
+
 (deftest notification-subscription-type-test
   (mt/with-temp [:model/Notification {n-id :id} {}]
     (testing "success path"

--- a/test/metabase/notification/payload/impl/card_test.clj
+++ b/test/metabase/notification/payload/impl/card_test.clj
@@ -576,14 +576,14 @@
           (testing "sanity: notification still exists but its payload record is gone"
             (is (t2/exists? :model/Notification (:id notification)))
             (is (not (t2/exists? :model/NotificationCard (:payload_id notification)))))
-          (testing "sending the orphaned notification should deactivate it rather than crash forever"
+          (testing "sending the orphaned notification should delete it rather than crash forever"
             (is (thrown-with-msg?
                  clojure.lang.ExceptionInfo
                  #"Card no longer exists"
                  (#'notification.send/send-notification-sync!
                   (t2/select-one :model/Notification (:id notification))))))
-          (testing "the notification should now be inactive"
-            (is (false? (t2/select-one-fn :active :model/Notification (:id notification))))))))))
+          (testing "the notification should be deleted"
+            (is (not (t2/exists? :model/Notification (:id notification))))))))))
 
 (defn- email->attachment-line-count
   [email]

--- a/test/metabase/notification/payload/impl/card_test.clj
+++ b/test/metabase/notification/payload/impl/card_test.clj
@@ -6,6 +6,7 @@
    [medley.core :as m]
    [metabase.channel.core :as channel]
    [metabase.notification.core :as notification]
+   [metabase.notification.models :as models.notification]
    [metabase.notification.payload.core :as notification.payload]
    [metabase.notification.payload.execute :as notification.payload.execute]
    [metabase.notification.send :as notification.send]
@@ -554,6 +555,35 @@
                 :task_details {:message (mt/malli=? [:fn #(str/includes? % "Division by zero")])}}]
               (t2/select [:model/TaskHistory :status :task_details] :task "notification-send"
                          {:order-by [[:started_at :asc]]}))))))
+
+(deftest orphaned-notification-deactivates-on-send-test
+  (testing "A notification whose card no longer exists should deactivate itself instead of crashing forever"
+    (notification.tu/with-notification-cleanup!
+      (mt/with-temp [:model/Card {card-id :id} {:name          notification.tu/default-card-name
+                                                :dataset_query (mt/mbql-query products {:aggregation [[:count]]})}]
+        (let [notification (models.notification/create-notification!
+                            {:payload_type :notification/card
+                             :payload      {:card_id card-id}
+                             :creator_id   (mt/user->id :crowberto)}
+                            []
+                            [@notification.tu/default-email-handler])]
+          (testing "sanity: notification is active"
+            (is (true? (t2/select-one-fn :active :model/Notification (:id notification)))))
+          ;; Simulate an orphaned notification: delete the NotificationCard directly.
+          ;; This mirrors what happens in prod when there is no Malli validation and a card is deleted (FK cascade deletes NotificationCard)
+          ;; but the Notification row itself survives.
+          (t2/delete! :model/NotificationCard (:payload_id notification))
+          (testing "sanity: notification still exists but its payload record is gone"
+            (is (t2/exists? :model/Notification (:id notification)))
+            (is (not (t2/exists? :model/NotificationCard (:payload_id notification)))))
+          (testing "sending the orphaned notification should deactivate it rather than crash forever"
+            (is (thrown-with-msg?
+                 clojure.lang.ExceptionInfo
+                 #"Card no longer exists"
+                 (#'notification.send/send-notification-sync!
+                  (t2/select-one :model/Notification (:id notification))))))
+          (testing "the notification should now be inactive"
+            (is (false? (t2/select-one-fn :active :model/Notification (:id notification))))))))))
 
 (defn- email->attachment-line-count
   [email]

